unpublished/build-system-abstraction-njs.rst

```diff
@@ -0,0 +1,708 @@
+    when absent. All tools reading the file MUST error on an unrecognised
+    schema version.
+
+bootstrap_requires
+    Optional list of PEP 508 dependency specifications that the
+    build frontend must ensure are available before invoking the build
+    backend. For instance, if using flit, then the requirements might
+    be::
+
+        "bootstrap_requires": ["flit"]
+
+build_backend
+    A mandatory string naming a Python object that will be used to
+    perform the build (see below for details). This is formatted
+    following the same ``module:object`` syntax as a ``setuptools``
+    entry point. For instance, if using flit, then the build system
+    might be specified as::
+
+        "build_system": "flit.api:main"
+
+    and this object would be looked up by executing the equivalent of::
+
+        import flit.api
+        backend = flit.api.main
+
+    It's also legal to leave out the ``:object`` part, e.g. ::
+
+        "build_system": "flit.api"
+
+    which acts like::
+
+        import flit.api
+        backend = flit.api
+
+    Formally, the string should satisfy this grammar::
+
+        identifier = (letter | '_') (letter | '_' | digit)*
+        module_path = identifier ('.' identifier)*
+        object_path = identifier ('.' identifier)*
+        entry_point = module_path (':' object_path)?
+
+    And we import ``module_path`` and then lookup
+    ``module_path.object_path`` (or just ``module_path`` if
+    ``object_path`` is missing).
+
+
+=========================
+ Build backend interface
+=========================
+
+The build backend object is expected to have attributes which provide
+some or all of the following hooks. The common ``config_settings``
+argument is described after the individual hooks::
+
+  def get_build_requires(config_settings):
+      ...
+
+This hook MUST return an additional list of strings containing PEP 508
+dependency specifications, above and beyond those specified in the
+``pypackage.json`` file. Example::
+
+  def get_build_requires(config_settings):
+      return ["wheel >= 0.25", "setuptools"]
+
+Optional. If not defined, the default implementation is equivalent to
+``return []``.
+
+::
+
+  def get_wheel_metadata(metadata_directory, config_settings):
+      ...
+
+Must create a ``.dist-info`` directory containing wheel metadata
+inside the specified ``metadata_directory`` (i.e., creates a directory
+like ``{metadata_directory}/{package}-{version}.dist-info/``. This
+directory MUST be a valid ``.dist-info`` directory as defined in the
+wheel specification, except that it need not contain ``RECORD`` or
+signatures. The hook MAY also create other files inside this
+directory, and a build frontend MUST ignore such files; the intention
+here is that in cases where the metadata depends on build-time
+decisions, the build backend may need to record these decisions in
+some convenient format for re-use by the actual wheel-building step.
```

Comment:
Yeah, unfortunately right now the name etc will be the same because there simply is no alternative -- e.g. these packages all satisfy a downstream requirement for `numpy`. This is one reason I'm keen to see the "reified extras" idea implemented, because among other things it would give us a framework for distinguishing these build variants.